File: generated_code/cpp/solution_3354_3.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

class Solution {
public:
    string minimizeStringValue(string s) {
        // Track the frequency of each lowercase letter in the string
        vector<int> counts(26, 0);
        int q_count = 0;
        
        for (char c : s) {
            if (c == '?') {
                q_count++;
            } else {
                counts[c - 'a']++;
            }
        }
        
        // Min-heap to always pick the character with the minimum frequency.
        // Pair structure: {frequency, character_index}
        // Greater comparison ensures we pick the smallest frequency, and 
        // lexicographically smallest character in case of ties.
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
        
        for (int i = 0; i < 26; ++i) {
            pq.push({counts[i], i});
        }
        
        // Collect the characters that will replace the '?' marks
        vector<char> replacements;
        replacements.reserve(q_count);
        
        for (int i = 0; i < q_count; ++i) {
            auto [freq, ch_idx] = pq.top();
            pq.pop();
            
            replacements.push_back('a' + ch_idx);
            
            // Increment frequency and push back into the heap
            pq.push({freq + 1, ch_idx});
        }
        
        // To make the final string lexicographically smallest, 
        // we must sort the replacement characters.
        sort(replacements.begin(), replacements.end());
        
        // Place the sorted replacements back into the '?' positions
        int rep_idx = 0;
        for (int i = 0; i < s.length(); ++i) {
            if (s[i] == '?') {
                s[i] = replacements[rep_idx++];
            }
        }
        
        return s;
    }
};

File: generated_code/python/solution_3354_3.py
from typing import List
import heapq

class Solution:
    def minimizeStringValue(self, s: str) -> str:
        # Count frequencies of existing characters in the string
        freq = [0] * 26
        question_count = 0
        for char in s:
            if char == '?':
                question_count += 1
            else:
                freq[ord(char) - ord('a')] += 1
        
        # If there are no '?' characters, return the string as is
        if question_count == 0:
            return s
        
        # Min-heap to choose characters that minimize the incremental cost.
        # Elements are tuples: (frequency, character_index)
        # This naturally prioritizes lower frequencies first, and breaks ties lexicographically.
        heap = [(freq[i], i) for i in range(26)]
        heapq.heapify(heap)
        
        # Determine which characters will replace the '?'s
        chosen_chars = []
        for _ in range(question_count):
            count, idx = heapq.heappop(heap)
            chosen_chars.append(chr(ord('a') + idx))
            # Push the updated frequency back into the heap
            heapq.heappush(heap, (count + 1, idx))
            
        # To make the final string lexicographically smallest, the chosen 
        # replacement characters must be sorted before inserting them back.
        chosen_chars.sort()
        
        # Reconstruct the string by replacing '?' with the sorted chosen characters
        result = []
        chosen_ptr = 0
        for char in s:
            if char == '?':
                result.append(chosen_chars[chosen_ptr])
                chosen_ptr += 1
            else:
                result.append(char)
                
        return "".join(result)